File: blazerod/render/main/physics/PhysicsScene.kt
package top.fifthlight.blazerod.physics

import org.joml.Vector3fc
import top.fifthlight.blazerod.model.RigidBody
import top.fifthlight.blazerod.model.util.MMD_SCALE
import top.fifthlight.blazerod.runtime.resource.RenderPhysicsJoint
import java.lang.ref.Reference
import java.nio.ByteBuffer
import java.nio.ByteOrder

class PhysicsScene(
    rigidBodies: List<RigidBody>,
    joints: List<RenderPhysicsJoint>,
) : AutoCloseable {
    private val pointer: Long
    private var closed = false
    internal val rigidBodyCount = rigidBodies.size

    init {
        require(rigidBodies.isNotEmpty()) { "Rigidbodies must not be empty" }

        if (!PhysicsLibrary.isPhysicsAvailable()) {
            throw IllegalStateException("Physics library is not available")
        }

        fun ByteBuffer.putVector3f(offset: Int, vector: Vector3fc) = this
            .putFloat(offset + 0, vector.x())
            .putFloat(offset + 4, vector.y())
            .putFloat(offset + 8, vector.z())

        val rigidBodyItemSize = 72
        val rigidBodiesBuffer = ByteBuffer.allocateDirect(rigidBodyItemSize * rigidBodies.size)
            .order(ByteOrder.nativeOrder())
        rigidBodies.forEachIndexed { index, rigidBody ->
            val offset = index * rigidBodyItemSize
            rigidBodiesBuffer.putInt(offset + 0, rigidBody.collisionGroup)
            rigidBodiesBuffer.putInt(offset + 4, rigidBody.collisionMask)
            rigidBodiesBuffer.putInt(offset + 8, rigidBody.shape.ordinal)
            rigidBodiesBuffer.putInt(offset + 12, rigidBody.physicsMode.ordinal)
            rigidBodiesBuffer.putVector3f(offset + 16, rigidBody.shapeSize)
            rigidBodiesBuffer.putVector3f(offset + 28, rigidBody.shapePosition)
            rigidBodiesBuffer.putVector3f(offset + 40, rigidBody.shapeRotation)
            rigidBodiesBuffer.putFloat(offset + 52, rigidBody.mass)
            rigidBodiesBuffer.putFloat(offset + 56, rigidBody.moveAttenuation)
            rigidBodiesBuffer.putFloat(offset + 60, rigidBody.rotationDamping)
            rigidBodiesBuffer.putFloat(offset + 64, rigidBody.repulsion)
            rigidBodiesBuffer.putFloat(offset + 68, rigidBody.frictionForce)
        }

        val jointItemSize = 108
        val jointsBuffer = ByteBuffer.allocateDirect(jointItemSize * joints.size)
            .order(ByteOrder.nativeOrder())
        joints.forEachIndexed { index, joint ->
            val offset = index * jointItemSize
            jointsBuffer.putInt(offset + 0, joint.type.ordinal)
            jointsBuffer.putInt(offset + 4, joint.rigidBodyAIndex)
            jointsBuffer.putInt(offset + 8, joint.rigidBodyBIndex)
            jointsBuffer.putVector3f(offset + 12, joint.position)
            jointsBuffer.putVector3f(offset + 24, joint.rotation)
            jointsBuffer.putVector3f(offset + 36, joint.positionMin)
            jointsBuffer.putVector3f(offset + 48, joint.positionMax)
            jointsBuffer.putVector3f(offset + 60, joint.rotationMin)
            jointsBuffer.putVector3f(offset + 72, joint.rotationMax)
            jointsBuffer.putVector3f(offset + 84, joint.positionSpring)
            jointsBuffer.putVector3f(offset + 96, joint.rotationSpring)
        }

        try {
            pointer = PhysicsLibrary.createPhysicsScene(rigidBodiesBuffer, jointsBuffer)
        } finally {
            Reference.reachabilityFence(rigidBodiesBuffer)
            Reference.reachabilityFence(jointsBuffer)
        }
    }

    internal fun getPointer(): Long {
        require(!closed) { "PhysicsScene is closed" }
        return pointer
    }

    override fun close() {
        if (closed) {
            return
        }
        PhysicsLibrary.destroyPhysicsScene(pointer)
        closed = true
    }
}

File: blazerod/render/common/physics/PhysicsScene.cpp
#include "PhysicsScene.h"

#include <sys/types.h>

#include <cstring>
#include <iostream>
#include <stdexcept>

namespace blazerod::physics {
template <typename T>
static void CopyField(T& field, void* data) {
    std::memcpy(&field, data, sizeof(field));
}

template <typename T>
static void CopyEnumField(T& field, void* data) {
    static_assert(std::is_enum_v<T>, "CopyEnumField only works with enum types");
    uint32_t enum_value;
    CopyField(enum_value, data);
    field = static_cast<T>(enum_value);
}

static void CopyVector3fField(Vector3f& vector, void* data) {
    uint8_t* ptr = static_cast<uint8_t*>(data);
    CopyField(vector.x, ptr + 0);
    CopyField(vector.y, ptr + 4);
    CopyField(vector.z, ptr + 8);
}

static std::vector<RigidBody> DeserializeRigidbodies(size_t size, void* data) {
    const size_t rigidbody_size = 72;
    if (size == 0) {
        throw std::invalid_argument("Empty rigidbody data");
    }
    if (size % rigidbody_size != 0) {
        throw std::invalid_argument("Invalid rigidbody size");
    }

    size_t rigidbody_count = size / rigidbody_size;
    std::vector<RigidBody> rigidbodies;
    rigidbodies.reserve(rigidbody_count);

    for (size_t i = 0; i < rigidbody_count; i++) {
        uint8_t* rigidbody_data = static_cast<uint8_t*>(data) + i * rigidbody_size;
        RigidBody rigidbody;

        CopyField(rigidbody.collision_group, rigidbody_data + 0);
        CopyField(rigidbody.collision_mask, rigidbody_data + 4);
        CopyEnumField(rigidbody.shape_type, rigidbody_data + 8);
        CopyEnumField(rigidbody.physics_mode, rigidbody_data + 12);
        CopyVector3fField(rigidbody.shape_size, rigidbody_data + 16);
        CopyVector3fField(rigidbody.shape_position, rigidbody_data + 28);
        CopyVector3fField(rigidbody.shape_rotation, rigidbody_data + 40);
        CopyField(rigidbody.mass, rigidbody_data + 52);
        CopyField(rigidbody.move_attenuation, rigidbody_data + 56);
        CopyField(rigidbody.rotation_damping, rigidbody_data + 60);
        CopyField(rigidbody.repulsion, rigidbody_data + 64);
        CopyField(rigidbody.friction_force, rigidbody_data + 68);

        rigidbodies.push_back(rigidbody);
    }

    return rigidbodies;
}

static std::vector<Joint> DeserializeJoints(size_t size, void* data) {
    const size_t joint_size = 108;
    if (size % joint_size != 0) {
        throw std::invalid_argument("Invalid joint size");
    }

    size_t joint_count = size / joint_size;
    std::vector<Joint> joints;
    joints.reserve(joint_count);

    for (size_t i = 0; i < joint_count; i++) {
        uint8_t* joint_data = static_cast<uint8_t*>(data) + i * joint_size;
        Joint joint;

        CopyEnumField(joint.type, joint_data + 0);
        CopyField(joint.rigidbody_a_index, joint_data + 4);
        CopyField(joint.rigidbody_b_index, joint_data + 8);
        CopyVector3fField(joint.position, joint_data + 12);
        CopyVector3fField(joint.rotation, joint_data + 24);
        CopyVector3fField(joint.position_min, joint_data + 36);
        CopyVector3fField(joint.position_max, joint_data + 48);
        CopyVector3fField(joint.rotation_min, joint_data + 60);
        CopyVector3fField(joint.rotation_max, joint_data + 72);
        CopyVector3fField(joint.position_spring, joint_data + 84);
        CopyVector3fField(joint.rotation_spring, joint_data + 96);

        joints.push_back(joint);
    }

    return joints;
}

PhysicsScene::PhysicsScene(size_t rigidbody_size, void* rigidbodies, size_t joints_size, void* joints)
    : rigidbodies(DeserializeRigidbodies(rigidbody_size, rigidbodies)),
      joints(DeserializeJoints(joints_size, joints)) {}
}  // namespace blazerod::physics
